src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/BufferSegment.cs

```diff
@@ -72,17 +72,23 @@ public void ResetMemory()
                 ArrayPool<byte>.Shared.Return(poolArray);
             }
 
-            // Order of below field clears is significant as it clears in a sequential order
-            // https://github.com/dotnet/corefx/pull/35256#issuecomment-462800477
-            Next = null;
             RunningIndex = 0;
             Memory = default;
             _memoryOwner = null;
-            _next = null;
             _end = 0;
             AvailableMemory = default;
         }
 
+        public void ResetSegmentLinks()
+        {
+            // Memory should have already been reset
+            Debug.Assert(_memoryOwner == null);
+            Debug.Assert(_end == 0);
+            Debug.Assert(RunningIndex == 0);
+
+            Next = null;
```

Comment:
Is this supposed to be the following?

```suggestion
            NextSegment = null;
```

That way `_next` gets reset as well.

Reply:
D'oh; yes, explains a few things....